File: pref_attach.go
// Generate directed graph using preferential attachment model
// described in the page 70 of the book "Complex Graphs and Networks"
// by Chung and Lu.

// Graph and vertex contain only the fields needed for the generation.

package main

import (
	"errors"
	"fmt"
	"math/rand"
	"os"

	
	"github.com/ajholanda/goadt"
)

const (
	INDEGREE = 0
	OUTDEGREE = 1
)

type Arc struct{
	tip *Vertex
	next *Arc
}

// Each vertex is used as a node in a doubly linked list sorted based
// on indegree or outdegree.
type Vertex struct{
	arcs *Arc
	degree [2]int // INDEGREE and OUTDEGREE values
}

type Graph struct{
	vertices []*Vertex // heads for lists sorted according to INDEGREE and OUTDEGREE
	n, m int // number of vertices and arcs
	totalDegree [2]int // total indegree and outdegree
	prioQueues [2]*adt.PriorityQueue // priority queues to store degree
}

func NewGraph(nvertices int) (*Graph) {
	var prioQueues [2]*adt.PriorityQueue 		

	for i := 0; i < 2; i++ {
		prioQueues[i] = adt.NewPriorityQueue(adt.MAX)
	}
	
	verts := make([]*Vertex, nvertices)	

	return &Graph{verts, 0, 0, [2]int{0, 0}, prioQueues}
}

func (G *Graph) NewVertex() (*Vertex){
	v := &Vertex{nil, [2]int{0,0}}
	G.n++

	return v
}

func (G *Graph) AddArc(u, v *Vertex) {	
	fmt.Println(u,v)

	for a := u.arcs; a != nil; a = a.next {
		// the arc already exists, return
		if a.tip == v {
			return
		}
	}
	fmt.Println("add arc", u, v)
	a := &Arc{v, u.arcs}
	u.arcs = a
	G.m++	
	//Stats and priority queues update
	u.degree[INDEGREE]++
	G.prioQueues[INDEGREE].Push(u.degree[INDEGREE], u, nil)
	
	v.degree[OUTDEGREE]++
	G.prioQueues[OUTDEGREE].Push(u.degree[OUTDEGREE], v, nil)
	
	G.totalDegree[INDEGREE]++	
	G.totalDegree[OUTDEGREE]++
}

func (G *Graph) requeue(typeDegree int) {
	for !G.prioQueues[typeDegree].IsEmpty() {// empty queue
		G.prioQueues[typeDegree].Pop()
	}

	for i := 0; i < G.n; i++ {
		v := G.vertices[i]

		G.prioQueues[typeDegree].Push((*v).degree[typeDegree], v, nil)
	}
}

func (G *Graph) RandomlyChooseVertexBy(typeDegree int) (*Vertex, error) {
	var t *Vertex
	p := rand.Float64() // // `0.0 <= f < 1.0`.
	var oldV *Vertex
	
	cumProb := 0.0
	oldV = nil
	N := 0
	for !G.prioQueues[typeDegree].IsEmpty() {
		h := G.prioQueues[typeDegree].Pop()
		v := h.Key().(*Vertex)
		
		deg := (*v).degree[typeDegree]

		fmt.Println(G, t, "\tN=", N, " deg=", deg)		

		if N == 0 && deg == 0 { // disconnected graph
			return v, nil;
		}
		
		cumProb += float64(deg) / float64(G.totalDegree[typeDegree])

		if cumProb > p { // the prob range was passed
			return oldV, nil
		}
		oldV = t
		N++
	}

	go G.requeue(typeDegree)
	
	return nil, errors.New("vertex can not be null")
}

// Add a new vertex v, and add a directed edge (u,v) from v by
// randomly and independently choosing u in proportion to the indegree
// of u in the current graph G.
func (G *Graph) SourceVertexStep() {
	v := G.NewVertex()

	if G.n < 2 { // not enough vertices to make arcs
		return
	}
	
	u, err := G.RandomlyChooseVertexBy(INDEGREE)

	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}
	G.AddArc(v, u)
}


// Add a new vertex v, and add a directed edge (u,v) to v by
// randomly and independently choosing u in proportion to the outdegree
// of u in the current graph G.
func (G *Graph) SinkVertexStep() {
	v := G.NewVertex()

	if G.n < 2 { // not enough vertices to make arcs
		return
	}
	
	u, err := G.RandomlyChooseVertexBy(OUTDEGREE)

	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}
	
	G.AddArc(u, v)
}


func main() {
	G := NewGraph(3)

	// Source-vertex-step
	G.SourceVertexStep()

	// Sink-vertex-step
	G.SinkVertexStep()

}

File: draft/pref_attach.pl
#!/usr/bin/env perl

# Generate directed graph using preferential attachment model
# described in the page 70 of the book "Complex Graphs and Networks"
# by Chung and Lu.

use strict;

my $TMPDIR = "/tmp/";
my $PROJ = "pref_attach";

my @deg_ts = ("in", "out"); # degree types

my $VERBOSE = 1;

my $SELF_EDGES = 0;
my $NSE = 0; # number of self edges

my $EPS = 0.00000000001; # to avoid division by zero 
my $EDGE = "->"; # edge link
my $NV = 0; # vertex counter
my %v2degree; # hash of hash, map vertex id to degree ( "in", "out")
my $total_degree = $EPS; # total degree = number of edges
my %G; # store edges and their weights

&main();

sub main{
    # we start with 1->2 edge to avoid void graph problems, division by
    # zero during first calculation of probabilities
    &add_edge(&init_vertex(), &init_vertex()); 

    #&ba(31, 36);

    &chung_lu(0.5, 0.2, .75, 0.0, 1170);
}

sub write_summary {
    print STDERR "Summary:\n";
    print STDERR "    #vertices: $NV\n";
    print STDERR "    #edges: $total_degree\n";
    print STDERR "    #total-degree(in|out): ".$total_degree."\n";
    print STDERR "    #self-edges: $NSE\n";
}

sub write_dot {
    my $fn = $TMPDIR.$PROJ.".dot";

    open(DOT, ">$fn") or die "Could not open $fn\n";
    print DOT "digraph g {\n";
    foreach my $e (keys %G) {
	my $w = $G{$e};
	$e =~ m/(\d+)$EDGE(\d+)/;
	my $u = $1;
	my $v = $2;

	print DOT "\t $u $EDGE $v [weight=$w];\n";
    }
    print DOT "}\n";
    close(DOT);
    print STDERR "Wrote $fn\n";
}

sub init_vertex {
    my $v = ++$NV;

    $v2degree{"in"}{$v} = $EPS;
    $v2degree{"out"}{$v} = $EPS;

    if ($VERBOSE) {
	my $ne = $total_degree;
	print STDERR "\tinit(v): $v\tindeg=".$v2degree{"in"}{$v}."/$ne\toutdeg=".$v2degree{"out"}{$v}."/$ne\n";
    }    
    return $v;
}

sub add_edge {
    my $u = shift;
    my $v = shift;
    my $e;
    
    if ($u < 0 || $v < 0) {
	die "ERROR: wrong index to vertex u=$u or v=$v\n";
    }

    if ($u == $v) {
	if (!$SELF_EDGES) {
	    return;
	} else {
	    $NSE++;
	}
    }

    $e = $u."->".$v;
    
    print STDERR "\tadd_edge: ".$e."\n";
    
    $v2degree{"out"}{$u}++;
    $v2degree{"in"}{$v}++;
    $total_degree++;

    if (exists($G{$e})) {
	$G{$e}++;
    } else {
	$G{$e} = 1;
    }
}

# Choose vertex in proportion to {in,out}degree.
sub choose_vertex {
    my $deg_type = shift; # in- or out-degree
    my $pr_plaw = shift; #  probability to occur power law distribution
    my $pr_uniform = shift; # probability to occur uniform distribution
    
    my $cum_prob = 0.0; # cummulative probabilty
    my $i = 0; # counter for the hash loop, i am not trusting in vertex id

    return if $NV == 0; # empty graph
    
    my $p_dist = rand(); # 
    
    my $p = rand();
    
    my $old_v = -1;
    if ($p_dist <= $pr_plaw) { # power law distribution
	foreach my $v (sort { $v2degree{$deg_type}{$b} <=> $v2degree{$deg_type}{$a} } keys %{ $v2degree{$deg_type} }) { # sort hash by value
	    $i++;
	    my $deg = $v2degree{$deg_type}{$v};
	    
	    
	    $cum_prob += $deg / $total_degree;	
	    print STDERR "\t POWER LAW choosing vertex: rand=$p\tv=$v/$NV\tdeg=$deg\tcum_prob=$cum_prob\n" if $VERBOSE;
	    
	    # when cumulative probability pass random number value, old v index
	    # is the vertex in the degree proportion. If there is only one element
	    # in the hash $old_v must be assigned to that element. We do that by
	    # verifying loop counter $i.
	    if ($cum_prob > $p && $i > 1) {
		last;
	    }	
	    $old_v = $v;
	} # END_foreach_my_$v
    } else { # uniform distribution
	foreach my $v ( keys %{ $v2degree{$deg_type} }) { # 
	    $i++;
	    $cum_prob = $i /$NV;
	    print STDERR "\t UNIFORM choosing vertex: rand=$p\tv=$v/$NV\ti=$i\tcum_prob=$cum_prob\n" if $VERBOSE;
	    if ($cum_prob > $p && $i > 1) {
		last;
	    }	
	    $old_v = $v;
	}
    }
    return $old_v;    
}

# Source-vertex-step - Add a new vertex v, and add a directed edge
# (v,w) from v by randomly and independently choosing w in proportion
# to the indegree of w in the current graph G.
sub source_vertex_step {
    my $pr_plaw = shift; #  probability to occur power law distribution
    my $pr_uniform = shift; # probability to occur uniform probability

    print STDERR "Source-vertex-step\n" if $VERBOSE;
    
    my $v = &init_vertex();

    my $w = &choose_vertex("in", $pr_plaw, $pr_uniform);

    &add_edge($v, $w, $pr_plaw, $pr_uniform);
}

# Sink-vertex-step - Add a new vertex v, and add a directed edge (u,v)
# to v by randomly and independently choosing u in proportion to the
# outdegree of u in the current graph G.
sub sink_vertex_step {
    print STDERR "Sink-vertex-step\n" if $VERBOSE;
    my $pr_plaw = shift; #  probability to occur power law distribution
    my $pr_uniform = shift; # probability to occur uniform probability
    
    my $v = &init_vertex();

    my $u = &choose_vertex("out", $pr_plaw, $pr_uniform);

    &add_edge($u, $v, $pr_plaw, $pr_uniform);
}

# Edge-step - Add a new edge (r,s) by indenpendently choosing vertices
# r and s with probability proportional to outdegree (respectively
# indegree).
sub edge_step {
    my $pr_plaw = shift; #  probability to occur power law distribution
    my $pr_uniform = shift; # probability to occur uniform probability
    print STDERR "Edge-step\n" if $VERBOSE;
    
    my $r = -1;
    my $s = -1;
    
    print STDERR "r>\n" if $VERBOSE;
    $r = &choose_vertex("out", $pr_plaw, $pr_uniform);

    print STDERR "s<\n" if $VERBOSE;    
    $s = &choose_vertex("in", $pr_plaw, $pr_uniform);

    &add_edge($r, $s, $pr_plaw, $pr_uniform);
}

# cumulative degree distributions
sub cum_degree_distrib {
    my %freq; # hash of hash ( "in|out" => degree => degree frequence )
    
    foreach my $t (@deg_ts) {
	my %freq;
	
	foreach my $v ( keys %{ $v2degree{$t} }) {
	    my $deg = $v2degree{$t}{$v};

	    if ($deg == $EPS) {
		next;
	    }
	    
	    if (exists($freq{$deg})) {
		$freq{$deg}++;
	    } else {
		$freq{$deg} = 1;
	    }
	}

	# write output
	my $cum_prob = 0.0;
	my $fn = "/tmp/$t-pref_attach.dat";
	open(OUT, ">$fn");	
	foreach my $deg (sort { $freq{$a} <=> $freq{$b} } keys %freq) {
	    $cum_prob += $freq{$deg} / $NV;

	    print STDERR "$deg\t".$freq{$deg}."\n";
	    print OUT "$deg\t$cum_prob\n";
	}
	close(OUT);
	print STDERR "Wrote $fn\n";
    }
}

# Taken from "Complex Graphs and Networks" book by Chung and Lu
sub chung_lu {
    my ($Pr1, $Pr2, $Pr_PL, $Pr_Unif, $MAX_NV) = @_; # p1, p2, Pr(power law), Pr(uniform), max number of vertices
    
    if ($Pr1 + $Pr2 > 1.0) {# p1 +p2 <=1
	die "ERROR: p1=$Pr1 + p2=$Pr2 must be <= 1.0\n";
    }

    if ($Pr_PL + $Pr_Unif > 1.0) {
	die "ERROR: Pr(Power Law Distribution)=$Pr1 + Pr(Uniform Distribution)=$Pr2 must be <= 1.0\n";
    }

    while (1) {
	# with probability p1, take source-vertex-step
	if ($Pr1 > rand()) {
	    &source_vertex_step($Pr_PL, $Pr_Unif);
	}

	# with probability p2, take sink-vertex-step
	if ($Pr2 > rand()) {
	    &sink_vertex_step($Pr_PL, $Pr_Unif);
	} else {# otherwise edge-step
	    &edge_step($Pr_PL, $Pr_Unif); 
	}
	
	last if $NV > $MAX_NV;
    }

    &write_summary();

    &cum_degree_distrib();

    &write_dot();

}

# Barabasi-Albert model
sub ba {
    my $nv = shift; # number of vertives
    my $ne = shift; # number of arcs/edges

    while ($NV<=$nv) {
	&source_vertex_step();
    }

    while ($total_degree<=$ne) {
	&edge_step();
    }

    &write_summary();

    &cum_degree_distrib();

    &write_dot();
}
